protos/flyteidl/artifact/artifacts.proto

```diff
@@ -0,0 +1,177 @@
+  // type to all Literals is a lot of work.
+  core.LiteralType type = 2;
+
+  // When you want to set tags on creation.
+  repeated Tag tags = 3;
+
+  // When you want to set an alias on creation. should this also be repeated?
+  repeated Alias aliases = 4;
+
+  oneof source {
+    // Outputs of tasks will have this.
+    core.TaskExecutionIdentifier task_execution = 5;
+
+    // Workflow outputs will have this.
+    core.WorkflowExecutionIdentifier execution = 6;
+
+    // Uploads, either from the UI or from the CLI, or FlyteRemote, will have this.
+    string principal = 7;
+  }
+
+  string short_description = 8;
+  string long_description = 9;
+}
+
+message ArtifactQuery {
+  core.ArtifactKey artifact_key = 1;
+  string version = 2;
+  Alias alias = 3;
+  repeated Tag tags = 4;
+}
+
+message CreateArtifactRequest {
+  core.ArtifactKey artifact_key = 1;
+
+  string version = 2; // optional
+
+  string uri = 3;
+
+  ArtifactSpec spec = 4;
+}
+
+message CreateArtifactResponse {
+  Artifact artifact = 1;
+}
+
+message GetArtifactRequest {
+  oneof identifier {
+    core.ArtifactID artifact_id = 1;
+
+    string uri = 2;
+  }
+
+  // If false, then long_description is not returned.
+  bool details = 3;
+}
+
+// Tags are general key/value pairs associated with an Artifact. They can be used to search and filter things.
+// Tags have to be unique by key for a given artifact. That is, for a given project/domain/name/version/key, there
+// can only be one value, obviously.
+message Tag {
+  // tag types can be user specified if they want, though we'd probably recommend sticking with some constants
+  // that we embed in various libraries.
+  string key = 1;
+  string value = 2;
+}
+
+// Add tags on an artifact.
+message AddTagsRequest {
+  core.ArtifactID artifact_id = 1;
+
+  // Why is this a repeated and not a map? Not sure. Just what ketan and I ended up on when we were quickly
+  // hacking this out.
+  repeated Tag tags = 2;
+}
+
+message AddTagsResponse {}
+
+// Probably will not do value match. Just remove by key.
+message RemoveTagsRequest {
+  core.ArtifactID artifact_id = 1;
+
+  repeated Tag tags = 2;  // can make these strings if necessary.
+}
+
+message RemoveTagsResponse {}
+
+message ListArtifactNamesRequest {
+  string project = 1;
+  string domain = 2;
+}
+
+message ListArtifactNamesResponse {
+  repeated core.ArtifactKey artifact_keys = 1;
+}
+
+message ListArtifactsRequest {
+  core.ArtifactKey artifact_key = 1;
+}
+
+message ListArtifactsResponse {
+  repeated Artifact artifacts = 1;
+}
+
+// Aliases identify a particular version of an artifact. They are different than tags in that they
+// have to be unique for a given artifact project/domain/name. That is, for a given project/domain/name/kind,
+// at most one version can have any given value at any point.
+message CreateAliasRequest {
+  core.ArtifactKey artifact_key = 1;
```

Comment:
Should this be ArtifactID since Alias is meant to identify a particular version of an artifact?